src/Functions/bech32.cpp

```diff
@@ -0,0 +1,488 @@
+    extern const int TOO_MANY_ARGUMENTS_FOR_FUNCTION;
+}
+
+/// Encode string to Bech32 or Bech32m address
+class EncodeToBech32Representation : public IFunction
+{
+public:
+    static constexpr auto name = "bech32Encode";
+
+    /// Default to the new and improved Bech32m algorithm
+    static constexpr int default_witver = 1;
+
+    static FunctionPtr create(ContextPtr) { return std::make_shared<EncodeToBech32Representation>(); }
+
+    String getName() const override { return name; }
+
+    bool isVariadic() const override { return true; }
+
+    size_t getNumberOfArguments() const override { return 0; }
+
+    bool isInjective(const ColumnsWithTypeAndName &) const override { return true; }
+
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const DataTypes & arguments) const override
+    {
+        if (arguments.size() < 2)
+            throw Exception(
+                ErrorCodes::TOO_FEW_ARGUMENTS_FOR_FUNCTION,
+                "At least two string arguments (hrp, data) are required for function {}",
+                getName());
+
+        if (arguments.size() > 3)
+            throw Exception(
+                ErrorCodes::TOO_MANY_ARGUMENTS_FOR_FUNCTION,
+                "A maximum of 3 arguments (hrp, data, witness version) are allowed for function {}",
+                getName());
+
+        /// check first two args, hrp and input string
+        for (size_t i = 0; i < 2; ++i)
+            if (!WhichDataType(arguments[i]).isStringOrFixedString())
+                throw Exception(
+                    ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                    "Illegal type {} of argument {} of function {}, expected String or FixedString",
+                    arguments[i]->getName(),
+                    i + 1,
+                    getName());
+
+        /// check 3rd (optional) arg, specifying witness version aka whether to use Bech32 or Bech32m algo
+        size_t arg_idx = 2;
+        if (arguments.size() == 3 && !WhichDataType(arguments[arg_idx]).isNativeUInt())
+            throw Exception(
+                ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT,
+                "Illegal type {} of argument {} of function {}, expected unsigned integer",
+                arguments[arg_idx]->getName(),
+                arg_idx + 1,
+                getName());
+
+        return std::make_shared<DataTypeString>();
+    }
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr &, size_t input_rows_count) const override
+    {
+        bool have_witver = arguments.size() == 3;
+
+        ColumnPtr col0 = arguments[0].column->convertToFullColumnIfConst();
+        ColumnPtr col1 = arguments[1].column->convertToFullColumnIfConst();
+        ColumnPtr col2;
+        if (have_witver)
+            col2 = arguments[2].column;
```

Comment:
The case when `witver` is a constant can be not working:
```
SELECT bech32Encode('bc', unhex(data), 1) FROM hex_data;
```

Reply:
I'm not sure what you mean, I have already included a test for a constant witver on 03512_bech32_functions.sql:63.
`SELECT bech32Encode(hrp, unhex(data), 1) FROM hex_data limit 1;`
This is slightly different from your test, but I think it tests the same thing, should I add the specific test you wrote?